pkg/webapi.go

```diff
@@ -0,0 +1,156 @@
+package giga
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/dogecoinfoundation/gigawallet/pkg/dogecoin"
+	"github.com/dogecoinfoundation/gigawallet/pkg/store"
+	"github.com/julienschmidt/httprouter"
+)
+
+// WebAPI implements tjstebbing/conductor.Service
+type WebAPI struct {
+	srv  *http.Server
+	port string
+	api  API
+}
+
+func NewWebAPI(config Config) (WebAPI, error) {
+	l1, err := dogecoin.NewL1Libdogecoin(config)
+	if err != nil {
+		return WebAPI{}, err
+	}
+	// TODO: this uses a mock store
+	api := NewAPI(store.NewMock(), l1)
+	return WebAPI{port: config.WebAPI.Port, api: api}, nil
+}
+
+func (t WebAPI) Run(started, stopped chan bool, stop chan context.Context) error {
+	go func() {
+		mux := httprouter.New()
+		mux.POST("/invoice/:foreignID", t.createInvoice)
+		mux.GET("/invoice/:invoiceID", t.getInvoice)
+		mux.POST("/account/:foreignID", t.createAccount)
+		mux.GET("/account/:foreignID", t.getAccount)
+		mux.GET("/account/byaddress/:address", t.getAccountByAddress) // TODO: figure out some way to to merge this and the above
+
+		t.srv = &http.Server{Addr: ":" + t.port, Handler: mux}
+		go func() {
+			if err := t.srv.ListenAndServe(); err != http.ErrServerClosed {
+				log.Fatalf("HTTP server ListenAndServe: %v", err)
+			}
+		}()
+		started <- true
+		select {
+		case ctx := <-stop:
+			// do some shutdown stuff then signal we're done
+			t.srv.Shutdown(ctx)
+			stopped <- true
+		}
+	}()
+	return nil
+}
+
+// createInvoice returns the ID of the created Invoice (which is the one-time address for this transaction) for the foreignID in the URL and the InvoiceCreateRequest in the body
+func (t WebAPI) createInvoice(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
+	foreignID := p.ByName("foreignID")
```

Comment:
I think we should be calling r.Context() and passing the ctx down into the t.api calls?

Reply:
Oh. I'm not too familiar with what contexts are useful for in Golang even though I've seen people use it (never have myself). Why are they useful here?